### PR TITLE
HRIS-249 [BE] Implement Add New Employee Functionality

### PR DESCRIPTION
**Pull request by arden-dave**
## Issue Link
https://framgiaph.backlog.com/view/HRIS-249

## Definition of Done
- [x] Admin user can add new employee
- [x] Has error handling and input validation
- [x] Created `Forms` table in the DB for initial employee form values
- [x] Generate initial employee form values upon adding new employee

## Notes
- Only Admin users can perform this function
- ❗ `IMPORTANT!` There should be `nextauth-token` in http header
- ⚠️ In the FE, there are no input fields for `Role` and `Schedule`

## Pre-condition
- run `dotnet ef database update` in the `api` directory
- (docker) run `docker compose up api db client --build -d`
- (local) run `dotnet run` in `api` directory
- go to `http://localhost:5257/graphql/`
- to add new employee, use the following mutation
```
mutation ($request: AddNewEmployeeRequestInput!) {
  addNewEmployee(request: $request)
}

# Variables
{
  "request": {
      "email": "test@test.com",
      "positionId" : 4,
      "scheduleId": 1,   
      "roleId" : 3,
      "firstName": "first",
      "middleName": "middle",
      "lastName": "last"
    }
}

# Don't forget the nextauth-token in the http header
```

## Expected Output
- Only Admin user should be able to perform this action
- There should be error message if any of the input are invalid
- Success mutation should return `true`
- On success, there should be new user in the `User` table in the DB
- On success, there should be also new data in `Forms` table

## Screenshots/Recordings

https://github.com/framgia/sph-hris/assets/111718037/073d86ea-5360-4c99-9694-b0f0d4482343

### CODE REVIEW

**Review by NiloJr-sun — APPROVED** [review]

File: api/Utils/Validation/EmployeeServiceInputValidation.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace api.Utils
+{
+    public class EmployeeServiceInputValidation : CustomInputValidation
+    {
+        public EmployeeServiceInputValidation(IDbContextFactory<HrisContext> contextFactory) : base(contextFactory)
+        {
+
+        }
```
Comment (abduljalilpalala) [inline api/Utils/Validation/EmployeeServiceInputValidation.cs:13]: Remove method if not needed.

File: api/Services/EmployeeService.cs [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+                    Type = FormTypeEnum.FIRST_DAY_ONBOARDING,
+                    Field = FirstDayOnboardingFieldEnum.GITHUB_ACCOUNT
+                },
+                new Form {
+                    User = user,
+                    Type = FormTypeEnum.FIRST_DAY_ONBOARDING,
+                    Field = FirstDayOnboardingFieldEnum.EMAIL_TWO_FACTOR_AUTH_SCREENSHOT
+                },
+                new Form {
+                    User = user,
```
Comment (abduljalilpalala) [inline api/Services/EmployeeService.cs:83]: Could you please try out this code? In this version, the method has been broken down into smaller methods, and the DRY principle has also been implemented.

```
private void GenerateEmployeeFormInitialData(User user, AddNewEmployeeRequest rawNewEmployeeData, HrisContext context)
{
    GenerateOnboardingForms(user, context);
    GeneratePersonalInformationForms(user, rawNewEmployeeData, context);
}

private void GenerateOnboardingForms(User user, HrisContext context)
{
    foreach (FirstDayOnboardingFieldEnum field in Enum.GetValues(typeof(FirstDayOnboardingFieldEnum)))
    {
        context.Forms.Add(new Form
        {
            User = user,
            Type = FormTypeEnum.FIRST_DAY_ONBOARDING,
            Field = field
        });
    }
}

private void GeneratePersonalInformationForms(User user, AddNewEmployeeRequest rawNewEmployeeData, HrisContext context)
{
    foreach (PersonalInformationFieldEnum field in Enum.GetValues(typeof(PersonalInformationFieldEnum)))
    {
        var form = new Form
        {
            User = user,
            Type = FormTypeEnum.PERSONAL_INFORMATION,
            Field = field
        };

        switch (field)
        {
            case PersonalInformationFieldEnum.FIRST_NAME:
                form.Value = rawNewEmployeeData.FirstName;
                break;
            case PersonalInformationFieldEnum.MIDDLE_NAME:
                form.Value = rawNewEmployeeData.MiddleName;
                break;
            case PersonalInformationFieldEnum.LAST_NAME:
                form.Value = rawNewEmployeeData.LastName;
                break;
            case PersonalInformationFieldEnum.POSITION:
                form.Value = user.Position.Name;
                break;
            default:
                break;
        }

        context.Forms.Add(form);
    }
}
```

File: api/Utils/Validation/EmployeeServiceInputValidation.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+            if (request.FirstName == "")
+                errors.Add(buildError(nameof(request.FirstName), InputValidationMessageEnum.INVALID_FIRST_NAME));
+
+            if (request.MiddleName != null && request.MiddleName == "")
+                errors.Add(buildError(nameof(request.MiddleName), InputValidationMessageEnum.INVALID_MIDDLE_NAME));
+
+            if (request.LastName == "")
+                errors.Add(buildError(nameof(request.LastName), InputValidationMessageEnum.INVALID_LAST_NAME));
+
```
Comment (abduljalilpalala) [inline api/Utils/Validation/EmployeeServiceInputValidation.cs:39]: Can you test this code for this section and see if it works?

```
if (string.IsNullOrEmpty(request.FirstName))
    {
        errors.Add(BuildError(nameof(request.FirstName), InputValidationMessageEnum.INVALID_FIRST_NAME));
    }

    if (string.IsNullOrEmpty(request.MiddleName))
    {
        errors.Add(BuildError(nameof(request.MiddleName), InputValidationMessageEnum.INVALID_MIDDLE_NAME));
    }

    if (string.IsNullOrEmpty(request.LastName))
    {
        errors.Add(BuildError(nameof(request.LastName), InputValidationMessageEnum.INVALID_LAST_NAME));
    }
```

File: api/Utils/Validation/EmployeeServiceInputValidation.cs (hunk 1 above)
Comment (arden-dave) [reply api/Utils/Validation/EmployeeServiceInputValidation.cs:13]: This block is a constructor because this class is inherited

**Review by jsvelte — APPROVED** [review]

**Pull request merged**